### ISSUE TITLE
[PROJECT TRACKER] `tsfresh` time-series features

### ISSUE THREAD

**Issue by topher-lo** [neocortexdb/functime#27]
This issue lists out the time-series features in [`tsfresh`](https://tsfresh.readthedocs.io/en/latest/text/list_of_features.html) and API design challanges and considerations.

❗ **IMPORTANT NOTE:** To ensure fair and useful benchmarks between our `polars` / Rust FFI code vs the original `numpy` code, our rewrite will go through three stages:
- [x] Rewrite using `polars` expressions only. Take note of features that require external dependencies. Modify original `tsfresh` API to improve user-experience.
- [x] Setup Rust FFI
- [x] Full benchmarking

## Proposed API Design
Each individual time-series feature should be implemented as a function, which takes a numeric type `pl.Expr` and returns another `pl.Expr` of any dtype (depends on the feature). Any parameter should be explicitly specified as function inputs. For example:

```python
import polars as pl

def some_feature(x: pl.Expr, lag: int) -> pl.Expr:
  ...
  return some_result
```

We will then register the functions into a custom `polars` namespace (see guide on "Extending the API" [here](https://pola-rs.github.io/polars/py-polars/html/reference/api.html)). So usage of `tsfresh` in `functime` will look like this:

```python
X_features = X.select([
  pl.col("value").ts.binned_entropy(n_bins=42),
  pl.col("value").ts.acf(lags=12),
  pl.col("value").ts.abs_energy()
])
```

This API design makes it easy to extend to panel data and rolling windows! For example:

```python
# Panel dataset
X_features = X.groupby("series_id").agg([
    pl.col("value").tsa.binned_entropy(n_bins=42),
    pl.col("value").tsa.acf(lags=12),
    pl.col("value").tsa.abs_energy()
])

# With window + stride (time-series)
X_features = X.groupby_dynamic(every="10i").agg([
    pl.col("value").tsa.binned_entropy(n_bins=42),
    pl.col("value").tsa.acf(lags=12),
    pl.col("value").tsa.abs_energy()
])

# With window + stride (panel)
X_features = X.groupby_dynamic(every="10i", by="series_id").agg([
    pl.col("value").tsa.binned_entropy(n_bins=42),
    pl.col("value").tsa.acf(lags=12),
    pl.col("value").tsa.abs_energy()
])
```

## Docstrings and Style
We use [`numpydoc`](https://numpydoc.readthedocs.io/en/latest/format.html) and [`black`](https://github.com/psf/black) formatting.

We also use [`pre-commit`](https://pre-commit.com/). After you clone the repo, remember to run the following:
```bash
pip install pre-commit
pre-commit install
```

## Features Checklist
If you want to work on a feature, just create a comment in this issue thread. I'll then add your @ to the checklist below.

❗ **IMPORTANT NOTE:** We are going to ignore any FFT (fast fourier transform) features for now!

- `abs_energy` @topher-lo 
- `abs_maximum` @topher-lo 
- `absolute_sum_of_energy` @topher-lo 
- ~`agg_linear_trend`~  Aggregated not supported.
- `approximate_entropy` @topher-lo 
- `ar_coeffs` @metaboulie 
- `augmented_dickrey_fuller`
- `autocorrelation` @achasol 
- `benfold_correlation` @MathieuCayssol
- `binned_entropy` @claysmyth 
- `c3` @mbignotti 
- `change_quantiles` @vienneraphael 
- `cid_ce` @mbignotti 
- `count_above` @achasol 
- `count_above_mean` @achasol 
- `count_below` @achasol 
- `count_below_mean` @achasol 
- `cwt_coefficients`
- `energy_ratio_by_chunks` @mbignotti 
- ~`fft_aggregated`~ Aggregated not supported.
- `fft_coefficient` (ignore)
- `first_location_of_maximum` @vienneraphael 
- `first_location_of_minimum` @vienneraphael 
- `fourier_entropy` @claysmyth 
- `friedrich_coefficients` @claysmyth 
- `has_duplicate` @achasol 
- `has_duplicate_max` @achasol 
- `has_duplicate_min` @achasol 
- `index_mass_quantile` @mbignotti 
- ~`kurtosis`~
- `large_std` @mbignotti 
- `last_location_of_maximum` @vienneraphael 
- `last_location_of_minimum` @vienneraphael 
- `lempel_ziv_complexity` @claysmyth 
- ~`length`~
- `linear_trend`
- ~`linear_trend_timewise`~ (not supported)
- `longest_strike_above_mean` @MathieuCayssol 
- `longest_strike_below_mean` @MathieuCayssol 
- `max_langevin_fixed_point` @claysmyth 
- ~`max`~
- ~`mean`~
- `mean_abs_change` @vienneraphael 
- `mean_change` @vienneraphael 
- `mean_n_abs_max` @MathieuCayssol 
- `mean_second_derivative_central`
- `median` (ignore...already in `polars`)
- `min` (ignore...already in `polars`)
- `number_crossing_m` @vienneraphael 
- `number_cwt_peaks` @MathieuCayssol 
- `number_peaks` @MathieuCayssol 
- `partial_autocorr` @claysmyth 
- `percentage_of_reoccuring_datapoints_to_all` @MathieuCayssol 
- `percentage_of_reoccuring_values_to_all` @MathieuCayssol 
- `permutation_entropy` @abstractqqq 
- ~`quantile`~
- ~`query_similarity_count`~ (not supported)
- `range_count` @abstractqqq 
- `ratio_beyond_r_sigma` @abstractqqq 
- `ratio_value_number_to_length` @abstractqqq 
- `root_mean_square` @abstractqqq 
- `sample_entropy` @abstractqqq 
- ~`skewness`~
- `spkt_welch_density` @abstractqqq 
- ~`std`~
- `sum_of_reoccuring_data_points`  (see `percentage_of_reoccuring_data_point_to_all`) @MathieuCayssol 
- `sum_of_reoccuring_values` (see `percentage_of_reoccuring_values_to_all`) @MathieuCayssol 
- ~`sum`~
- `symmetry_looking` @metaboulie 
- `time_reversal_asymmetry_statistic` @metaboulie 
- ~`value_count`~
- ~`variance`~
- `var_greater_than_std` @vienneraphael 
- `var_coeff` @mbignotti 

## Testing
White-box testing strategy. We use `pytest`. Each developer is responsible for their own unit test per feature. Should be relatively straightforward: can reuse most code from `tsfresh` [unit tests](https://github.com/blue-yonder/tsfresh/blob/main/tests/units/feature_extraction/test_feature_calculations.py)

## Development Workflow
Everybody create ONE pull request with all your implemented features and tests. The title of the PR should be `feat: tsfresh features @your-github-handle`

## Challenges Checklist
Here is a list of challenges I discovered during a first attempt at the [rewrite](https://github.com/descendant-ai/functime/pull/25). Please add to this list in the comments below!

1. Some features in `tsfresh` do not list of parameters in the function signature (e.g. [`cwt_coefficients`](https://tsfresh.readthedocs.io/en/latest/_modules/tsfresh/feature_extraction/feature_calculators.html#cwt_coefficients)). We should explicitly list it out for our implementation. You can find the default parameters from the original implementation [here](https://github.com/blue-yonder/tsfresh/blob/main/tsfresh/feature_extraction/settings.py).
2. Some features in `tsfresh` return a pandas `Series` (e.g. [`ar_coefficients`](https://tsfresh.readthedocs.io/en/latest/api/tsfresh.feature_extraction.html#tsfresh.feature_extraction.feature_calculators.ar_coefficient)). For our implementation, we should return a `pl.List` instead.
3. We should ignore `agg_` features. Let the user aggregate using polars expressions e.g. `x.list.to_explode().std()` instead.

**Comment by topher-lo** [comment]
- `abs_energy`
- `abs_maximum`
- `absolute_sum_of_energy`
- `agg_linear_trend`
- `approximate_entropy`

**Comment by MathieuCayssol** [comment]
- `benfold_correlation`
- `longest_strike_above_mean `
- `longest_strike_below_mean `
- `mean_n_abs_max`
- `percentage_of_reoccuring_datapoints_to_all`
- `percentage_of_reoccuring_values_to_all`
- `sum_of_reoccuring_data_points`
- `sum_of_reoccuring_values`
- `number_peaks` (not yet implemented)
- `number_cwt_peaks` (not yet implemented)

**Comment by abstractqqq** [comment]
kurtosis is already in Polars.

permutation_entropy
query_similarity_count
range_count
ratio_beyond_r_sigma
ratio_value_number_to_length
root_mean_square
sample_entropy
spkt_welch_density

**Comment by topher-lo** [comment]
> * spkt_welch_density ( Is this fourier related?)

@abstractqqq Not FFT, but it does use the `welch` algorithm (from `scipy`) for computing spectral density: https://github.com/scipy/scipy/blob/v1.11.2/scipy/signal/_spectral_py.py#L297-L460. You could try use using [`welch`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.welch.html) for now. We can consider rewriting it in Rust down the line.

**Comment by achasol** [comment]
- autocorrelation

- count_above
- count_above_mean

- count_below
- count_below_mean 

- has_duplicate
- has_duplicate_max
- has_duplicate_min

**Comment by claysmyth** [comment]
- binned_entropy
- fourier_entropy
- friedrich_coefficients
- lempel_ziv_complexity
- max_langevin_fixed_point
- partial_autocorr

**Comment by mbignotti** [comment]
- `c3`
- `cid_ce`
- `energy_ration_by_chunks`
- `index_mass_quantile`
- `large_std`
- `var_coeff`

**Comment by topher-lo** [comment]
Stable version released with v0.9.0!
